lib/charms/postgresql_k8s/v0/postgresql_tls.py

```diff
@@ -0,0 +1,181 @@
+        """Manager of PostgreSQL relation with TLS Certificates Operator."""
+        super().__init__(charm, "client-relations")
+        self.charm = charm
+        self.peer_relation = peer_relation
+        self.certs = TLSCertificatesRequiresV1(self.charm, TLS_RELATION)
+        self.framework.observe(
+            self.charm.on.set_tls_private_key_action, self._on_set_tls_private_key
+        )
+        self.framework.observe(
+            self.charm.on[TLS_RELATION].relation_joined, self._on_tls_relation_joined
+        )
+        self.framework.observe(
+            self.charm.on[TLS_RELATION].relation_broken, self._on_tls_relation_broken
+        )
+        self.framework.observe(self.certs.on.certificate_available, self._on_certificate_available)
+        self.framework.observe(self.certs.on.certificate_expiring, self._on_certificate_expiring)
+
+    def _on_set_tls_private_key(self, event: ActionEvent) -> None:
+        """Set the TLS private key, which will be used for requesting the certificate."""
+        self._request_certificate(event.params.get("private-key", None))
+
+    def _request_certificate(self, param: Optional[str]):
+        """Request a certificate to TLS Certificates Operator."""
+        if param is None:
+            key = generate_private_key()
+        else:
+            key = self._parse_tls_file(param)
+
+        csr = generate_csr(
+            private_key=key,
+            subject=self.charm.get_hostname_by_unit(self.charm.unit.name),
+            sans=self._get_sans(),
+            additional_critical_extensions=self._get_tls_extensions(),
+        )
+
+        self.charm.set_secret(SCOPE, "key", key.decode("utf-8"))
+        self.charm.set_secret(SCOPE, "csr", csr.decode("utf-8"))
+
+        if self.charm.model.get_relation(TLS_RELATION):
+            self.certs.request_certificate_creation(certificate_signing_request=csr)
+
+    @staticmethod
+    def _parse_tls_file(raw_content: str) -> bytes:
+        """Parse TLS files from both plain text or base64 format."""
+        if re.match(r"(-+(BEGIN|END) [A-Z ]+-+)", raw_content):
+            return re.sub(
+                r"(-+(BEGIN|END) [A-Z ]+-+)",
+                "\\1",
+                raw_content,
+            ).encode("utf-8")
+        return base64.b64decode(raw_content)
+
+    def _on_tls_relation_joined(self, _) -> None:
+        """Request certificate when TLS relation joined."""
+        self._request_certificate(None)
+
+    def _on_tls_relation_broken(self, _) -> None:
+        """Disable TLS when TLS relation broken."""
+        self.charm.set_secret(SCOPE, "ca", None)
+        self.charm.set_secret(SCOPE, "cert", None)
+        self.charm.set_secret(SCOPE, "chain", None)
+        self.charm.update_config()
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        """Enable TLS when TLS certificate available."""
+        if event.certificate_signing_request != self.charm.get_secret(SCOPE, "csr"):
```

Comment:
I'm not sure if this is addressed here, but this check failed for me in the past as the `tls_certificates` lib `str.strip()`'s the CSR that we generate, and returned the `strip()`'d CSR back. Maybe your check handles it though.

Reply:
In this case, the right CSR content was being returned.